analysis/tensorflow/fast_em.py

```diff
@@ -0,0 +1,169 @@
+#!/usr/bin/python
+"""
+fast_em.py: Tensorflow implementation of expectation maximization for RAPPOR
+association analysis.
+
+TODO:
+  - Use TensorFlow ops for reading input (so that reading input can be
+    distributed)
+  - Reduce the number of ops (currently proportional to the number of reports).
+    May require new TensorFlow ops.
+"""
+
+import sys
+
+import numpy as np
+import tensorflow as tf
+
+
+def log(msg, *args):
+  if args:
+    msg = msg % args
+  print >>sys.stderr, msg
+
+
+def ExpectTag(f, expected):
+  """Read and consume a 4 byte tag from the given file."""
+  b = f.read(4)
+  if b != expected:
+    raise RuntimeError('Expected %r, got %r' % (expected, b))
+
+
+def ReadListOfMatrices(f):
+  """
+  Read a big list of conditional probability matrices from a binary file.
+  """
+  ExpectTag(f, 'ne \0')
+  num_entries = np.fromfile(f, np.uint32, count=1)[0]
+  log('Number of entries: %d', num_entries)
+
+  ExpectTag(f, 'es \0')
+  entry_size = np.fromfile(f, np.uint32, count=1)[0]
+  log('Entry size: %d', entry_size)
+
+  ExpectTag(f, 'dat\0')
+  vec_length = num_entries * entry_size
+  v = np.fromfile(f, np.float64, count=vec_length)
+
+  log('Values read: %d', len(v))
+  log('v: %s', v[:10])
+  #print 'SUM', sum(v)
+
+  # NOTE: We're not reshaping because we're using one TensorFlow tensor object
+  # per matrix, since it makes the algorithm expressible with current
+  # TensorFlow ops.
+  #v = v.reshape((num_entries, entry_size))
+
+  return num_entries, entry_size, v
+
+
+def WriteTag(f, tag):
+  if len(tag) != 3:
+    raise AssertionError("Tags should be 3 bytes.  Got %r" % tag)
+  f.write(tag + '\0')  # NUL terminated
+
+
+def WriteResult(f, num_em_iters, pij):
+  WriteTag(f, 'emi')
+  emi = np.array([num_em_iters], np.uint32)
+  emi.tofile(f)
+
+  WriteTag(f, 'pij')
+  pij.tofile(f)
+
+
+def DebugSum(num_entries, entry_size, v):
+  """Sum the entries as a sanity check."""
+  cond_prob = tf.placeholder(tf.float64, shape=(num_entries * entry_size,))
+  debug_sum = tf.reduce_sum(cond_prob)
+  with tf.Session() as sess:
+    s = sess.run(debug_sum, feed_dict={cond_prob: v})
+  log('Debug sum: %f', s)
+
+
+def ExpectationMaximization(num_entries, entry_size, v, max_em_iters,
+                            epsilon=1e-6):
+  """EM algorithm (using the TensorFlow API).
+
+  Args:
+    num_entries: number of matrices (one per report)
+    entry_size: total number of cells in each matrix
+    v: numpy.ndarray (e.g. 7000 x 8 matrix)
+    max_em_iters: maximum number of EM iterations
+
+  Returns:
+    pij: numpy.ndarray (e.g. vector of length 8)
+  """
+  # Placeholder for the value from the previous iteration.
+  pij_in = tf.placeholder(tf.float64, shape=(entry_size,))
+
+  # split along dimension 0
+  v_split = tf.split(0, num_entries, v)
```

Comment:
See if there is a better way to do this. Either through a different OP as Olfar suggested or by reusing a constant v_split.